### PR TITLE
Adding a notification service with a SNS implementation

### PR DESCRIPTION
**Pull request by tgianos**
Implementations receive internal events and can publish external notifications based on implementations of the interfaces.

### CODE REVIEW

File: metacat-common/src/main/java/com/netflix/metacat/common/MetacatRequestContext.java [hunk 1]
```diff
@@ -32,6 +35,9 @@
     public static final String HEADER_KEY_JOB_ID = "X-Netflix.job.id";
     public static final String HEADER_KEY_DATA_TYPE_CONTEXT = "X-Netflix.data.type.context";
 
+    private final String id = UUID.randomUUID().toString();
+    // TODO: Move to Java 8 and use java.time.Instant
+    private final long timestamp = new Date().getTime();
```
Comment (ajoymajumdar) [inline metacat-common/src/main/java/com/netflix/metacat/common/MetacatRequestContext.java]: How about using Joda Instant?

File: metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageType.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * @author tgianos
+ * @since 0.1.47
+ */
+public enum SNSMessageType {
+    TABLE_CREATE,
+    TABLE_DELETE,
+    TABLE_UPDATE,
+    TABLE_PARTITIONS_UPDATE,
+    PARTITION_CREATE,
```
Comment (ajoymajumdar) [inline metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageType.java]: PARTITION_ADD

File: metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageType.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * Enumeration of the various types of SNS events there can be.
+ *
+ * @author tgianos
+ * @since 0.1.47
+ */
+public enum SNSMessageType {
+    TABLE_CREATE,
+    TABLE_DELETE,
+    TABLE_UPDATE,
+    TABLE_PARTITIONS_UPDATE,
```
Comment (ajoymajumdar) [inline metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageType.java]: Do we track partition add and delete at the table level?

File: metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageFactory.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+     * Convert a JSON String into a message if possible.
+     *
+     * @param json The body of the message to convert back to the original message object from JSON string
+     * @return The message bound back into a POJO
+     * @throws IOException When the input isn't valid JSON
+     */
+    public SNSMessage<?> getMessage(final String json) throws IOException {
+        final JsonNode object = this.mapper.readTree(json);
+        if (object.has(TYPE_FIELD)) {
+            final SNSMessageType messageType = SNSMessageType.valueOf(object.get(TYPE_FIELD).asText());
```
Comment (ajoymajumdar) [inline metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageFactory.java]: Is there a need for sub classes of SNSMessage?

File: metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageType.java (hunk 3 above)
Comment (tgianos) [inline metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageType.java]: Is this a rhetorical question? The event that comes in from internal has a qname and only one qname for the entire payload of partitions so I assume that is the qname of the table they're for... no? If that's not the case we probably need a whole other big feature outside the scope of this PR to fix that before we can send the notification of a table changing. Unless you're saying we need to parse each of the partitions to figure out what table they belong to?

File: metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageFactory.java (hunk 4 above)
Comment (tgianos) [inline metacat-common/src/main/java/com/netflix/metacat/common/dto/notifications/sns/SNSMessageFactory.java]: I made the sub classes mostly out of convenience for bindings and usage by clients. It allows them to do instance of and not have complicated payload bindings to deal with as the types are already set. Functionally they're just wrappers as you noticed but just there for convenience.

File: metacat-main/src/main/java/com/netflix/metacat/main/services/notifications/sns/SNSNotificationsModule.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+ */
+@Slf4j
+public class SNSNotificationsModule extends AbstractModule {
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void configure() {
+        final Multibinder<NotificationService> notificationServices = Multibinder.newSetBinder(
```
Comment (ajoymajumdar) [inline metacat-main/src/main/java/com/netflix/metacat/main/services/notifications/sns/SNSNotificationsModule.java]: Nice.

File: metacat-main/src/main/java/com/netflix/metacat/main/services/notifications/sns/SNSNotificationsModule.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * If SNS is enabled this will return an instance of the SNS Notification service implementation.
+     *
+     * @param config       The metacat configuration
+     * @param objectMapper The JSON object mapper to use
+     * @return If enabled a SNS Notification Service implementation. Otherwise null.
+     */
+    @Provides
+    public NotificationService getSnsNotificationService(
```
Comment (ajoymajumdar) [inline metacat-main/src/main/java/com/netflix/metacat/main/services/notifications/sns/SNSNotificationsModule.java]: Where is this provider used?

Comment (tgianos) [inline metacat-main/src/main/java/com/netflix/metacat/main/services/notifications/sns/SNSNotificationsModule.java]: Whoops it's not. I forgot to delete it after I moved it into its own class. Good catch will remove

**Pull request merged**